File: 2024/Day 5/Part 1/Part 1.cpp
#include <iostream>
#include <vector>
#include <unordered_map>
#include <unordered_set>
#include <sstream>
#include <string>
#include <algorithm>
#include <fstream>

using namespace std;

bool isCorrectOrder(const vector<int>& update, const unordered_map<int, unordered_set<int>>& rules) {
    unordered_map<int, int> positions;
    for (int i = 0; i < update.size(); ++i) {
        positions[update[i]] = i;
    }
    for (const auto& rule : rules) {
        int before = rule.first;
        for (int after : rule.second) {
            if (positions.count(before) && positions.count(after) && positions[before] > positions[after]) {
                return false;
            }
        }
    }
    return true;
}

int main() {
    ifstream inputFile("C:\\Users\\Nuño\\Advent of Code\\2024\\Day 5\\input.txt");
    if (!inputFile.is_open()) {
        cerr << "Failed to open the file." << endl;
        return 1;
    }

    string line;
    unordered_map<int, unordered_set<int>> rules;
    vector<vector<int>> updates;
    bool readingRules = true;

    while (getline(inputFile, line)) {
        if (line.empty()) {
            readingRules = false;
            continue;
        }
        stringstream ss(line);
        if (readingRules) {
            int x, y;
            char sep;
            ss >> x >> sep >> y;
            rules[x].insert(y);
        } else {
            vector<int> update;
            int page;
            while (ss >> page) {
                if (ss.peek() == ',') ss.ignore();
                update.push_back(page);
            }
            updates.push_back(update);
        }
    }

    inputFile.close();

    int sum = 0;
    for (const auto& update : updates) {
        if (isCorrectOrder(update, rules)) {
            sum += update[update.size() / 2];
        }
    }

    cout << "Sum: " << sum << endl;
    return 0;
}

File: 2024/Day 4/Part 1/Part 1.py
def find_word(grid, word):
    def search(x, y, dx, dy):
        for k in range(len(word)):
            if not (0 <= x + k * dx < len(grid) and 0 <= y + k * dy < len(grid[0])):
                return False
            if grid[x + k * dx][y + k * dy] != word[k]:
                return False
        return True

    directions = [(1, 0), (0, 1), (1, 1), (1, -1), (-1, 0), (0, -1), (-1, -1), (-1, 1)]
    count = 0
    for i in range(len(grid)):
        for j in range(len(grid[0])):
            for dx, dy in directions:
                if search(i, j, dx, dy):
                    count += 1
    return count

file = open("C:\\Users\\Nuño\\Advent of Code\\2024\\Day 4\\input.txt")
lines = file.read().split("\n")
count = 0

grid = [list(line) for line in lines]

print(find_word(grid, "XMAS"))
file.close()